stdlib/2.7/__builtin__.pyi

```diff
@@ -210,7 +212,7 @@ class unicode(basestring, Sequence[unicode]):
     def center(self, width: int, fillchar: unicode = u' ') -> unicode: ...
     def count(self, x: unicode) -> int: ...
     def decode(self, encoding: unicode = ..., errors: unicode = ...) -> unicode: ...
-    def encode(self, encoding: unicode = ..., errors: unicode = ...) -> str: ...
+    def encode(self, encoding: unicode = ..., errors: unicode = ...) -> bytes: ...
```

Comment:
Should things like `encoding` have the type `Text` or `NativeStr` (or even `str`)? Consider various cases:

1. `encoding: Text`. If we have `'x' -> NativeStr` then `u'foo'.encode('utf8')` would be a type error, since `NativeStr` is not compatible with `Text`.
2. `encoding: NativeStr`. Then `u'foo'.encode(u'utf8')` would result in a false positive.
3. `encoding: str`. Then we really don't have any type checks, any string-like object will be OK.

The idea of adding `NativeStr` was that in PY2 it represents cases where an implicit unicode->bytes ASCII conversion takes place. And this is the case for `encoding` in `unicode.encode()`.

But having to write `u'foo'.encode(NativeStr(u'utf8'))` is not that convenient. (Here the ASCII types proposal would have worked better, but only when you use literals).

On the other hand, if you have your encoding name as a string parameter (as opposed to a literal, and maybe it's a more important case overall) and you make `encoding` in `unicode.encode()` to be of type `NativeStr` (as its run-time semantics suggests), then the cases are like that:

1. `def f(encoding: Text): u'foo'.encode(encoding)`. You have to use `u'foo'.encode(encoding.encode('ascii') if PY2 else encoding)`. Here you have to make your implicit conversion explicit by saying that you believe that this text can always be ASCII-converted to a 8-bit string. This way you make your code PY2+3 compatible.
2. `def f(encoding: NativeStr): u'foo'.encode(encoding)`. Everything looks good here.
3. `def f(encoding: str): u'foo'.encode(encoding)`. Again everything is OK, but it looks like another way of saying that you are sure that you are ASCII-compatible here.

So, a possible conclusion from this is: if you are sure that your argument gets implicitly ASCII-converted to a 8-bit string, then you should use `NativeStr` for it. If you want to state that you don't really care about checking for implicit ASCII conversions (and PY2+3 compatibility in general), you should use `str`.